calico/common.py

```diff
@@ -601,105 +608,132 @@ def _validate_rules(rules_dict, issues):
-                    issues.append("ICMP type is not an integer in rule %s." %
-                                  rule)
-                elif not 0 <= icmp_type <= 255:
-                    issues.append("ICMP type is out of range in rule %s." %
-                                  rule)
-            icmp_code = rule.get("icmp_code")
-            if icmp_code is not None:
-                if not isinstance(icmp_code, int):
-                    issues.append("ICMP code is not an integer in rule %s." %
-                                  rule)
-                elif not 0 <= icmp_code <= 255:
-                    issues.append("ICMP code is out of range.")
-                if icmp_type is None:
-                    # ICMP code without ICMP type not supported by iptables;
-                    # firewall against that.
-                    issues.append("ICMP code specified without ICMP type.")
+            for neg_pfx in ("", "!"):
+                _validate_rule_match_criteria(rule, issues, neg_pfx)
 
             unknown_keys = set(rule.keys()) - KNOWN_RULE_KEYS
             if unknown_keys:
                 issues.append("Rule contains unknown keys: %s." % unknown_keys)
 
 
+def _validate_rule_match_criteria(rule, issues, neg_pfx):
+    """Validates and canonicalises a rule's match criteria.
+
+    Each call validates either the negated or non-negated match criteria.
+    I.e. the ones with "!" prefixed or not.
+
+    :param rule: The dict for the individual rule.  If this contains selectors,
+           they are replaced with parsed versions.  Protocols and IPs are
+           also normalised.
+    :param list[str] issues: List of issues found.  This method appends any
+           issues it finds to the list.
+    :param str neg_pfx: The negation prefix, "" for positive matches or "!"
+           for negative.
+    """
+    # Absolutely all fields are optional, but some have valid and
+    # invalid values.
+    assert neg_pfx in ("", "!")
+
+    # Explicitly get the non-negated protocol; even negated matches on port
+    # or ICMP values require the protocol to be specified.
+    protocol = rule.get("protocol")
+
+    # Check the (possibly negated) profile.
+    protocol_key = neg_pfx + 'protocol'
+    maybe_neg_proto = rule.get(protocol_key)
+    if maybe_neg_proto is not None and maybe_neg_proto not in KERNEL_PROTOCOLS:
+        issues.append("Invalid %s %s in rule %s" %
+                      (protocol_key, maybe_neg_proto, rule))
+    elif maybe_neg_proto is not None:
+        maybe_neg_proto = intern(str(maybe_neg_proto))
+        rule[protocol_key] = str(maybe_neg_proto)
+
+    ip_version = rule.get('ip_version')
+    if ip_version == 4 and protocol == "icmpv6":
+        issues.append("Using icmpv6 with IPv4 in rule %s." % rule)
+    if ip_version == 6 and protocol == "icmp":
+        issues.append("Using icmp with IPv6 in rule %s." % rule)
+
+    for tag_type in (neg_pfx + 'src_tag', neg_pfx + 'dst_tag'):
+        tag = rule.get(tag_type)
```

Comment:
Match criteria in the same rule are anded together so `{"!src_tag": "foo", "src_tag": "bar"}` matches IPs that belong to endpoints in tag `bar`, unless they're in tag `foo`.

`{"!src_tag": "foo", "src_tag": "foo"}` wouldn't match anything.